src/formats/SDF.cpp

```diff
@@ -0,0 +1,258 @@
+
+                property_value = file_->readline();
+            } else {
+                // Continuation of a property value
+                property_value += '\n';
+                property_value += line;
+            }
+        } catch (const FileError& e) {
+            warning("Premature end of SDF File during global property reading!");
+            return;
+        }
+    }
+
+}
+
+void SDFFormat::write(const Frame& frame) {
+    auto& topology = frame.topology();
+    auto& positions = frame.positions();
+    assert(frame.size() == topology.size());
+
+    if (frame.get("name")) {
+        fmt::print(*file_, "{}\n", frame.get("name")->as_string());
+    } else {
+        fmt::print(*file_, "NONAME\n");
+    }
+    fmt::print(*file_, " chemfiles-lib\n\n");
+    fmt::print(*file_, "{:>3}{:>3}  0     0  0  0  0  0  0999 V2000\n", frame.size(), topology.bonds().size());
+
+    for (size_t i = 0; i < frame.size(); i++) {
+        std::string type = topology[i].type();
+
+        if (type == "" || type.length() > 3) {
+            type = "Xxx";
+        }
+
+        fmt::print(
+            *file_, "{:>10.4f}{:>10.4f}{:>10.4f} {:3} 0  0  0  0  0  0  0  0  0  0  0  0\n",
+            positions[i][0], positions[i][1], positions[i][2], type
+        );
+    }
+
+    for (const auto& bond : topology.bonds()) {
+        fmt::print(
+            *file_, "{:>3}{:>3}  1  0  0  0  0\n",
+            bond[0] + 1, bond[1] + 1
+        );
+    }
+
+    fmt::print(*file_, "M END\n$$$$\n");
+    steps_positions_.push_back(file_->tellg());
+}
+
+bool forward(TextFile& file) {
+    if (!file) {return false;}
+
+    long long natoms = 0;
+    long long nbonds = 0;
+    try {
+        // Ignore junk lines
+        file.readline();
+        file.readline();
+        file.readline();
+        std::string counts_line = file.readline();
+
+        if (counts_line.length() < 10) {
+            throw format_error("Counts line must have at least 10 digits, it has {}", counts_line.length());
+        }
+
+        natoms = std::stoll(counts_line.substr(0,3));
+        nbonds = std::stoll(counts_line.substr(3,3));
+    } catch (const FileError&) {
+        // No more line left in the file
+        return false;
+    } catch (const std::invalid_argument&) {
+        // We could not read an integer, so give up here
+        return false;
+    }
+
+    if (natoms < 0 || nbonds < 0) {
+        throw format_error(
+            "number of atoms and bonds can not be negative in '{}'", file.filename()
+        );
+    }
+
+    try {
+        file.readlines(static_cast<size_t>(natoms) + static_cast<size_t>(nbonds));
+    } catch (const FileError&) {
+        // We could not read the lines from the file
+        throw format_error(
+            "not enough lines in '{}' for SDF format", file.filename()
+        );
+    }
+
+    // Search for ending character, updating the position in the file only
+    while (!file.eof()) {
+        if (file.readline() == "$$$$") break;
```

Comment:
Should this condition also check for `M END`, as `read`?

Reply:
No, the only delimiter between compounds is `$$$$` (this is the difference between SDF and MOL as MOL is single compound only).

 `M  END` denotes the end of the atom property block (which I don't use). Although the format guarantees the presence of `M  END`, some older programs don't write it, so I am being safe and only checking for `$$$$`, which MUST be present if the program supports the SDF format, otherwise, something went really wrong with writing the file.

Reply:
OK, thanks! Does this means that the SDF reader can open MOL files ?

Reply:
Technically yes, that's why the `forward` function returns true at the end. A `MOL` file would just end, but `SDF` requires the presence of `$$$$` at the end of the file. In fact (to my understanding), `SDF` is just a wrapper around `MOL`/`MDL` to support more properties and multiple compounds.